### PR TITLE
Copybutton

### LINKED ISSUE #86: Simpler way to distribute player links
Make it easier to distribute the links to the players:
* After each links: a button to copy the link to the clipboard
* Maybe: A button to copy the whole link table to the clipboard such that it can easily pasted into a chat. (Assuming you can trust the players such that they only use their own links :-P ).

### PR DESCRIPTION
**Pull request by matthewejones**
Created copy buttons next to the links for each player.

Also created a "copy all" button which copies all links in a formatted way.

Resolves #86

### CODE REVIEW

File: src/client/pages/create.js [hunk 1]
```diff
@@ -225,11 +242,18 @@ class Create extends React.Component {
                 <td>
                   <a href={`${window.location.origin}/${this.state.gameID}/${i}/${this.state.secret[i]}`} target="_blank" rel="noopener noreferrer">{window.location.origin}/{this.state.gameID}/{i}/{this.state.secret[i]}</a>
                 </td>
+                <td>
+                  <Button onClick={() => this.copyToClipboard(`${window.location.origin}/${this.state.gameID}/${i}/${this.state.secret[i]}`)}>
```
Comment (ChristophNiehoff) [inline src/client/pages/create.js:246]: As this URL is created at 3 different places, could you refactor it to a function `i => url(i)` ?

File: src/client/pages/create.js [hunk 2]
```diff
@@ -136,6 +138,21 @@ class Create extends React.Component {
     })
   }
 
+  async copyToClipboard(text) {
+    //Not sure I fully understand binding, should this function be bound to this?
```
Comment (ChristophNiehoff) [inline src/client/pages/create.js:142]: I found this nice summary: https://gist.github.com/fongandrew/f28245920a41788e084d77877e65f22f

As far as I understand it, in this particular case you don't need to bind as you are using an arrow function in line 246.

File: src/client/pages/create.js [hunk 3]
```diff
@@ -136,6 +138,21 @@ class Create extends React.Component {
...
+    return await navigator.clipboard.writeText(text);
+  }
+
+  copyAllLinks() {
+    this.copyToClipboard(
+      "You have been invited to a game of Elevation of Privilege:\n" +
+      Array(this.state.players).fill(0).map((v, i) => {
+        return `${this.state.names[i]}:\t${window.location.origin}/${this.state.gameID}/${i}/${this.state.secret[i]}\n`;
+      }).join("")
```
Comment (ChristophNiehoff) [inline src/client/pages/create.js:151]: Would it make it more readable to remove the `\n` from the previous line and join via `.join("\n")` ?

File: src/client/pages/create.js (hunk 1 above)
Comment (ChristophNiehoff) [reply src/client/pages/create.js:246]: Idea:  Would it look nicer if we didn't use a button but rather an icon like https://materialui.co/icon/content-copy ?

Comment (ChristophNiehoff) [reply src/client/pages/create.js:246]: And yet another idea: I think on other websites there usually is an animation when copying to the clipboard (like an appearing check mark or so). I believe from a UX perspective this is done to give feedback to the user that something did indeed happen.  Can we include this easily?

File: src/client/pages/create.js [hunk 4]
```diff
@@ -225,11 +242,18 @@ class Create extends React.Component {
...
+                  <Button onClick={() => this.copyToClipboard(`${window.location.origin}/${this.state.gameID}/${i}/${this.state.secret[i]}`)}>
+                    Copy
+                  </Button>
+                </td>
               </tr>
             )}
             </tbody>
           </Table>
           <hr />
+          <Button onClick={this.copyAllLinks} color="warning" block size="lg">Copy All</Button>
```
Comment (ChristophNiehoff) [inline src/client/pages/create.js:255]: Just a general comment:  Is there some way to write a test for this?

**Review by ChristophNiehoff — CHANGES_REQUESTED** [review]

**Comment by ChristophNiehoff** [comment]
Works like charm!

File: src/client/pages/create.js (hunk 3 above)
Comment (ChristophNiehoff) [reply src/client/pages/create.js:151]: maybe even join with 2 newlines? Then it's easier for the players to copy their link from the chat?

File: src/client/pages/create.js (hunk 2 above)
Comment (matthewejones) [reply src/client/pages/create.js:142]: Thanks, this was really useful

File: src/client/pages/create.js (hunk 3 above)
Comment (matthewejones) [reply src/client/pages/create.js:151]: Good idea

File: src/client/pages/create.js (hunk 1 above)
Comment (matthewejones) [reply src/client/pages/create.js]: I've refactored the url generation into a function.

I agree that an icon might look a bit neater and the tick's probably a good idea to say whether copying was successful

File: src/client/components/copybutton/copybutton.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    render() {
+        return (
+            <Button color={this.state.color} onClick={() => this.handleClick()}>
+                Copy
+            </Button>
+        )
+    }
+
+    handleClick() {
+        copyToClipboard(this.state.text).then(() => {
```
Comment (ChristophNiehoff) [inline src/client/components/copybutton/copybutton.js:25]: `this.state.text` is not defined. Therefore, not the link but the string `undefined` gets copied.

File: src/client/components/copybutton/copybutton.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    constructor(props) {
+        super(props);
+        this.state = {color: this.props.color};
+    }
+
+    render() {
+        return (
+            <Button color={this.state.color} onClick={() => this.handleClick()}>
```
Comment (ChristophNiehoff) [inline src/client/components/copybutton/copybutton.js:18]: Usually one would bind `handleClick` to `this` and pass the onClick-handler as `<Button onClick={this.handleClick} />`. Could you change it to that pattern?

File: src/client/pages/create.js [hunk 7]
```diff
@@ -223,13 +239,18 @@ class Create extends React.Component {
               <tr key={i}>
                 <td>{this.state.names[i]}</td>
                 <td>
-                  <a href={`${window.location.origin}/${this.state.gameID}/${i}/${this.state.secret[i]}`} target="_blank" rel="noopener noreferrer">{window.location.origin}/{this.state.gameID}/{i}/{this.state.secret[i]}</a>
+                  <a href={`${this.url(i)}`} target="_blank" rel="noopener noreferrer">{window.location.origin}/{this.state.gameID}/{i}/{this.state.secret[i]}</a>
```
Comment (ChristophNiehoff) [inline src/client/pages/create.js:242]: We should also use `url` in this line.

File: src/client/pages/create.js (hunk 4 above)
Comment (ChristophNiehoff) [reply src/client/pages/create.js:252]: Could you also implement the color effect for this button?

File: src/utils/utils.js [hunk 8]
```diff
@@ -86,4 +86,10 @@ export function getTypeString(type) {
     return map[type];
   }
   return "";
+}
+
+//I thought because it doesn't necessarily only apply to create.js it might belong here
```
Comment (ChristophNiehoff) [inline src/utils/utils.js:91]: Great idea!

File: src/client/pages/create.js [hunk 9]
```diff
@@ -223,13 +239,18 @@ class Create extends React.Component {
...
+                </td>
+                <td>
+                  <CopyButton text={this.url(i)} />
```
Comment (ChristophNiehoff) [inline src/client/pages/create.js:245]: You don't pass the color that you marked as required in `<CopyButton/>`?

**Review by ChristophNiehoff — CHANGES_REQUESTED** [review]

**Comment by ChristophNiehoff** [comment]
@matthewejones I really like the color effect!

File: src/client/components/copybutton/copybutton.js (hunk 5 above)
Comment (matthewejones) [reply src/client/components/copybutton/copybutton.js:25]: Whoops, I meant text to be a property

File: src/client/pages/create.js (hunk 7 above)
Comment (matthewejones) [reply src/client/pages/create.js:242]: fixed it and did a search to make sure there weren't any others that I missed

File: src/client/pages/create.js (hunk 9 above)
Comment (matthewejones) [reply src/client/pages/create.js:245]: I didn't mean to make it required, I just wanted it to be passable, I've fixed this

File: src/client/components/copybutton/copybutton.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+class CopyButton extends React.Component {
+    static propTypes = {
+        text: PropTypes.string.isRequired,
+        color: PropTypes.string,
+        active: PropTypes.bool,
+        block: PropTypes.bool,
+        disabled: PropTypes.bool,
+        outline: PropTypes.bool,
+        size: PropTypes.string
```
Comment (matthewejones) [inline src/client/components/copybutton/copybutton.js:16]: Maybe there's a better way to pass these properties to the Button

**Comment by ChristophNiehoff** [comment]
@matthewejones I'm afraid long player names break the layout...

Screenshot from 2021-07-15 15-03-15

**Comment by ChristophNiehoff** [comment]
> @matthewejones I'm afraid long player names break the layout...
But I think this is not a problem of your changes. This can be resolved in https://github.com/dehydr8/elevation-of-privilege/issues/94

**Comment by matthewejones** [comment]
> > @matthewejones I'm afraid long player names break the layout...
> 
> But I think this is not a problem of your changes. This can be resolved in #94

Ok, maybe I can have a look at that tomorrow

File: src/client/pages/create.js (hunk 4 above)
Comment (matthewejones) [reply src/client/pages/create.js]: In terms of testing, I can't work out an easy way to test the stuff I've added

Comment (ChristophNiehoff) [reply src/client/pages/create.js]: One way might be to make `copyToClipboard` a property in `<CopyButton />` (see "dependency injection").
Then, in a test, you could pass in a mock and see whether it is called with the correct arguments.

**Review by ChristophNiehoff — APPROVED** [review]

**Comment by ChristophNiehoff** [comment]
I think, this really looks awesome!  From my side this can be merged.

**Review by ChristophNiehoff — APPROVED** [review]

**Pull request merged**